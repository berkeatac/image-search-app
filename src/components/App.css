body {
  margin: 0;
}

.navbar {
  position: sticky;
  top: 0;
  z-index: 20;
}

.content {
  max-width: 960px;
  margin: 30px auto;
}

.loader-container {
  display: flex;
  justify-content: center;
}
