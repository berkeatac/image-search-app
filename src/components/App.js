import React from "react";

const App = () => {
  return <div>Image Search App</div>;
};

export default App;
