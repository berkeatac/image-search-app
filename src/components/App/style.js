import styled from "styled-components";

export const AppContent = styled.div`
  max-width: 960px;
  margin: 30px auto;
`;

export const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
`;
