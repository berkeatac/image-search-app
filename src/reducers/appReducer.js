export const initialState = {
  images: [],
  term: "istanbul",
  collection: 0,
  page: 1,
  total_pages: 1,
};

export const appReducer = (state, action) => {
  switch (action.type) {
    case "SET_PARAMS":
      return {
        ...state,
        term: action.payload.term,
        collection: action.payload.collection,
        page: 1,
      };
    case "SET_IMAGES":
      return {
        ...state,
        images: action.payload.images,
        total_pages: action.payload.total_pages,
      };
    case "SET_PAGE":
      return {
        ...state,
        page: action.payload.page,
      };
    default:
      return state;
  }
};
