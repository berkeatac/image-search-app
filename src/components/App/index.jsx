import React, { useState, useEffect, useReducer } from "react";
import ClipLoader from "react-spinners/ClipLoader";
import { createGlobalStyle } from "styled-components";

import Header from "../Header";
import ImageList from "../ImageList";
import PageButtons from "../PageButtons";
import ErrorBox from "../ErrorBox";
import Footer from "../Footer";
import imageGetter from "../../api/imageGetter";
import { appReducer, initialState } from "../../reducers/appReducer";

import { AppContent, LoaderContainer } from "./style";

const GlobalStyle = createGlobalStyle`
  body {
    margin: 0px;
  }
`;

const App = () => {
  const [state, dispatch] = useReducer(appReducer, initialState);
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);

  const onSearchSubmit = async () => {
    setError("");
    setLoading(true);

    const [data, error] = await imageGetter(
      state.term,
      state.collection,
      state.page,
      setError
    );
    if (error) {
      setError(error);
    } else {
      const { results, total_pages } = data;
      dispatch({
        type: "SET_IMAGES",
        payload: { images: results, total_pages },
      });
      if (!results.length) setError("No result has been found");
    }

    setLoading(false);
  };

  useEffect(() => {
    onSearchSubmit();
  }, [state.term, state.collection, state.page]);

  return (
    <div>
      <GlobalStyle />
      <Header state={state} dispatch={dispatch} />
      <AppContent>
        {loading && (
          <LoaderContainer className="loader-container">
            <ClipLoader />
          </LoaderContainer>
        )}
        {error !== "" && !loading && <ErrorBox errorMessage={error} />}
        {!loading && error === "" && (
          <>
            <ImageList images={state.images} />
            <PageButtons state={state} dispatch={dispatch} />
          </>
        )}
      </AppContent>
      <Footer />
    </div>
  );
};

export default App;
