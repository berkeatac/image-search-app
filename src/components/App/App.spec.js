import React from "react";
import renderer from "react-test-renderer";

import App from "./index";
import appReducer from "../../reducers/appReducer";

describe("App", () => {
  describe("Reducer", () => {
    it("should set page with reducer", () => {
      const state = {
        images: [],
        term: "istanbul",
        collection: 0,
        page: 1,
        total_pages: 1,
      };
      const newState = appReducer(state, {
        type: "SET_PAGE",
        payload: { page: 2 },
      });

      expect(newState).toEqual({
        images: [],
        term: "istanbul",
        collection: 0,
        page: 2,
        total_pages: 1,
      });
    });

    it("should set page to 1 with new params", () => {
      const state = {
        images: [],
        term: "istanbul",
        collection: 0,
        page: 5,
        total_pages: 1,
      };
      const newState = appReducer(state, {
        type: "SET_PARAMS",
        payload: { term: "ankara", collection: 0 },
      });

      expect(newState).toEqual({
        images: [],
        term: "ankara",
        collection: 0,
        page: 1,
        total_pages: 1,
      });
    });
  });

  test("snapshot renders", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
