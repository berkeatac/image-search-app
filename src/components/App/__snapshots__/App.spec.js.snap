// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div>
  <div
    className="sc-AxjAm ddUaJj"
  >
    <div
      className="sc-AxirZ fxSvMx"
    >
      <div
        className="sc-AxiKw gjZAwv"
      >
        <img
          alt="logo"
          className="sc-AxhCb efvFGU"
          src="logo.svg"
        />
      </div>
    </div>
    <form
      className="sc-AxhUy knSQqY"
      onSubmit={[Function]}
    >
      <input
        className="sc-AxgMl cJivbW"
        data-cy="query-text"
        placeholder="Query"
        type="text"
      />
      <div
        className="sc-AxheI cNckAa css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-a97y1b-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1uiylfs-ValueContainer"
          >
            <div
              className=" css-1uccc91-singleValue"
            >
              Default
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-2-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-62wm9a-IndicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-pgs6m7-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <img
                alt="dropdown arrow"
                src="dropdown_indicator.svg"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="sc-AxmLO gdRRGW"
        data-cy="query-button"
        type="submit"
      >
        <p
          className="sc-Axmtr gBPFzr"
        >
          SEARCH
        </p>
      </button>
    </form>
  </div>
  <div
    className="sc-fzokOt ka-dUud"
  >
    <div
      className="sc-fzozJi gNforj"
      data-cy="image-list"
    />
    <div
      className="sc-fzpans hqXyBQ"
    >
      <button
        className="sc-fznyAO kjDhaP"
        data-cy="prev-page"
        disabled={true}
        onClick={[Function]}
      >
        <p
          className="sc-fzplWN iZNOQC"
        >
          Previous
        </p>
      </button>
      <button
        className="sc-fznyAO kjDhaP"
        data-cy="next-page"
        disabled={true}
        onClick={[Function]}
      >
        <p
          className="sc-fzplWN iZNOQC"
        >
          Next
        </p>
      </button>
    </div>
  </div>
</div>
`;
